### PR TITLE
backend/overlay: add async support

### PR DESCRIPTION
**Pull request by bpowers**
hi @jvilk,

Consider this a proposal for async support in OverlayFS.  A couple of
big things here:

- with the lower directory being async, operations (like rename,
  rmdir, and others) may ping-pong back and forth between requests to
  the upper layer and lower layer, to check if files exist or
  copy/delete directory contents.  I'm interested in stacking multiple
  overlays on top of each other, which will only exacerbate the
  situation.  To deal with this, I've added Java-style locking to the
  exposed OverlayFS object -- only one request can proceed at a time,
  to avoid thinking about correctness when you interleave two fs
  requests together.  I need to go back and see if there is more
  locking that should happen around the File object.
- async versions of all the methods on OverlayFS.
- a few small fixes, such as using the buffer size if a Content-Length
  header isn't returned on an XMLHttpRequest (such as for chunked
  responses?)

I'm still doing more testing with browsix - so I'd like to hold off
until I extend some of browsix's tests to make sure nothing is hinkey,
but so far it is looking good!

### CODE REVIEW

File: src/backend/AsyncMirror.ts [hunk 1]
```diff
@@ -106,7 +107,7 @@ export default class AsyncMirror extends file_system.SynchronousFileSystem imple
                 if (err) {
                   cb(err);
                 } else if (i < files.length) {
-                  copyItem(`${p}/${files[i]}`, copyNextFile);
```
Comment (jvilk) [inline src/backend/AsyncMirror.ts]: Did this line cause an error at one point? (I avoided `join` because it's more expensive)

File: src/backend/OverlayFS.ts [hunk 2]
```diff
@@ -2,34 +2,55 @@ import file_system = require('../core/file_system');
...
+import {LockedFS} from '../generic/locked_fs';
 import path = require('path');
 let deletionLogPath = '/.deletedFiles.log';
 
 /**
  * Given a read-only mode, makes it writable.
  */
 function makeModeWritable(mode: number): number {
-  return 0x92 | mode;
+  return 0o222 | mode;
```
Comment (jvilk) [inline src/backend/OverlayFS.ts]: Nice; I didn't know you could do octals like this. I would love to see these replacing all of the hex values across the codebase.

File: src/backend/OverlayFS.ts [hunk 3]
```diff
@@ -75,6 +93,43 @@ export default class OverlayFS extends file_system.SynchronousFileSystem impleme
     };
   }
 
+  private createParentDirectoriesAsync(p: string, cb: ()=>void): void {
```
Comment (jvilk) [inline src/backend/OverlayFS.ts]: Please make this function pass an optional error to the callback.

File: src/backend/OverlayFS.ts [hunk 4]
```diff
@@ -75,6 +93,43 @@ export default class OverlayFS extends file_system.SynchronousFileSystem impleme
...
+
+      let dir = toCreate.pop();
+      _this._readable.stat(dir, false, (err: ApiError, stats?: Stats) => {
+        // stop if we couldn't read the dir
+        if (!stats)
+          return cb();
+
+        // FIXME: get correct permissions
+        _this._writable.mkdir(dir, stats.mode, (err?: ApiError) => {
+          // FIXME: better error handling
```
Comment (jvilk) [inline src/backend/OverlayFS.ts]: Forward error through the callback.

File: src/backend/OverlayFS.ts [hunk 5]
```diff
@@ -96,9 +151,15 @@ export default class OverlayFS extends file_system.SynchronousFileSystem impleme
     return true;
   }
 
-  public _syncSync(file: preload_file.PreloadFile<any>): void {
+  public _syncAsync(file: PreloadFile<UnlockedOverlayFS>, cb: (err: ApiError)=>void): void {
+    this.createParentDirectoriesAsync(file.getPath(), () => {
```
Comment (jvilk) [inline src/backend/OverlayFS.ts]: Once you add it, check error argument from `createParentDirectoriesAsync` before proceeding.

File: src/backend/OverlayFS.ts [hunk 6]
```diff
@@ -75,6 +93,43 @@ export default class OverlayFS extends file_system.SynchronousFileSystem impleme
...
+        toCreate.push(parent);
+        parent = path.dirname(parent);
+        _this._writable.stat(parent, false, statDone);
+      } else {
+        createParents();
+      }
+    }
+
+    function createParents(): void {
+      if (!toCreate.length)
```
Comment (jvilk) [inline src/backend/OverlayFS.ts]: (Style nit) Please use braces around the `if` body.

File: src/backend/OverlayFS.ts [hunk 7]
```diff
@@ -117,40 +178,37 @@ export default class OverlayFS extends file_system.SynchronousFileSystem impleme
...
-            0x1a4, (err: ApiError, fd?: file.File) => {
-            if (err) {
-              end(err);
-            } else {
-              this._deleteLog = fd;
-              end();
-            }
-          });
+    // if we're already initialized, immediately invoke the callback
+    if (this._isInitialized)
```
Comment (jvilk) [inline src/backend/OverlayFS.ts]: (Style nit) Please use braces around the `if` body.

Comment (jvilk) [inline src/backend/OverlayFS.ts]: (Will stop adding these nits in PR, as they appear in multiple places.)

File: src/generic/locked_fs.ts [hunk 8]
```diff
@@ -0,0 +1,392 @@
+import {Mutex} from './mutex';
```
Comment (jvilk) [inline src/generic/locked_fs.ts]: What is this `LockedFS`? It looks incomplete.

File: src/generic/mutex.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
+
+declare var setImmediate: (cb: Function) => void;
+
+/// non-recursive mutex
+export class Mutex {
```
Comment (jvilk) [inline src/generic/mutex.ts]: When you have a file w/ one export, do `export default`.

File: src/generic/locked_fs.ts (hunk 8 above)
Comment (jvilk) [inline src/generic/locked_fs.ts]: Ah I see; `OverlayFS` uses it now.

**Comment by jvilk** [comment]
It looks like you didn't modify the unit tests to construct an asynchronous OverlayFS. Please do that so the tests cover your changes.

**Comment by jvilk** [comment]
You just need to construct a second `OverlayFS` instance here that is asynchronous:

https://github.com/jvilk/BrowserFS/blob/master/test/harness/factories/overlay_factory.ts

Also, if there are any crazy corner cases that are important to test, you can add OverlayFS-specific tests here:

https://github.com/jvilk/BrowserFS/tree/master/test/tests/fs/OverlayFS

File: src/backend/AsyncMirror.ts (hunk 1 above)
Comment (bpowers) [inline src/backend/AsyncMirror.ts]: Yes -- without it directories or files in the root directory give paths like `"//file"` (two slashes), and `copyItem` passes this directly to the bottom filesystem -- XHRFS in my tests.  XHRFS has names like `'/file'` in its map of files, not `'//file'`.  I could add a conditional here to check if `p[p.length-1] === '/'` to avoid calling join, although it looks somewhat reasonable on posix at least: https://github.com/nodejs/node/blob/master/lib/path.js#L1207

File: src/backend/OverlayFS.ts (hunk 2 above)
Comment (bpowers) [inline src/backend/OverlayFS.ts]: cool, I'll work on that in a followup PR

Comment (jvilk) [inline src/backend/OverlayFS.ts]: Good, good.

File: src/backend/OverlayFS.ts (hunk 3 above)
Comment (bpowers) [inline src/backend/OverlayFS.ts]: done

**Comment by bpowers** [comment]
@jvilk I've addressed (I think) all of your comments except for the IE8 failure, and your comment about test generation.  In fact, I think the problem is the reverse: most of the existing functionality for OverlayFS was being tested through the `SynchronousFileSystem` methods, i.e.:

```js
  public stat(p: string, isLstat: boolean, cb: Function): void {
    try {
      cb(null, this.statSync(p, isLstat));
    } catch (e) {
      cb(e);
    }
  }
```

Now that `OverlayFS` has these directly, the original functionality isn't getting tested.  Not sure the best thing to do here, I guess I could split the sync and async parts into separate classes, and have the async overlayFS inherit from the sync one so they can be tested independently?  Thats probably the cleanest thing actually.

**Comment by jvilk** [comment]
Good point. I consider that a bug in our test suite -- if our test suite isn't covering synchronous paths through the file systems, then that needs to be remedied separately.

File: src/generic/locked_fs.ts [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+*/
+
+/// This class serializes access to an underlying async filesystem.
+/// For example, on an OverlayFS instance with an async lower
+/// directory operations like rename and rmdir may involve multiple
+/// requests involving both the upper and lower filesystems -- they
+/// are not executed in a single atomic step.  OverlayFS uses this
+/// LockedFS to avoid having to reason about the correctness of
+/// multiple requests interleaving.
+export class LockedFS<T extends FileSystem> implements FileSystem {
```
Comment (jvilk) [inline src/generic/locked_fs.ts]: Make this `export default`.

File: src/generic/locked_fs.ts [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+  private _fs: T;
+  private _mu: Mutex;
+
+  constructor(cls: {new(...args: any[]): T}, ...args: any[]) {
+    this._fs = new (Function.prototype.bind.apply(cls, [null].concat(args)));
```
Comment (jvilk) [inline src/generic/locked_fs.ts]: `Function.prototype.bind` is not supported in IE8. Please add a polyfill to our polyfill file.

Comment (jvilk) [inline src/generic/locked_fs.ts]: I'm not completely following the logic here, either; shouldn't you pass it an instantiated `FileSystem` object? Why call it dynamically here?

**Comment by jvilk** [comment]
My only other question is if I should wait for `LockedFile` to land before merging this.

**Comment by bpowers** [comment]
@jvilk might as well wait for LockedFile, I'll do that today.  Also, the most recent appveyor fail seems spurious.

**Comment by jvilk** [comment]
Re-running; it was spurious. I need to have it cache `node_modules`.

**Comment by bpowers** [comment]
@jvilk another spurious failure for appveyor.  I changed my mind: if its ok with you, I'd like to merge without LockedFile.  I think LockedFile is "more correct", but all my testing so far has been without it, and I haven't seen issues.
